File: db_manager.py
# ملف: db_manager.py
import motor.motor_asyncio
import time

# ======================
# إعداد MongoDB
# ======================
MONGO_URI = "mongodb+srv://wpee923_db_user:08520852KR@cluster0.nzjd5gc.mongodb.net/?retryWrites=true&w=majority"
DB_NAME = "imperial_bot"

client = motor.motor_asyncio.AsyncIOMotorClient(MONGO_URI)
db = client[DB_NAME]
users_col = db["users"]
groups_col = db["groups"]

# ======================
# المطورين
# ======================
DEVELOPERS = [5860391324, 7076215547, 7855813063]

# ======================
# المستخدمين
# ======================
async def get_user(uid, name=None, username=None):
    user = await users_col.find_one({"uid": uid})
    if not user:
        user = {
            "uid": uid,
            "gold": 0,
            "bank": 0,
            "points": 0,
            "rank": 0,
            "last_gift_time": 0,
            "total_messages": 0,
            "level": 0,
            "name": name,
            "username": username,
            "married_to": None,
            "birthday": None,
            "birthday_auto": True
        }
        await users_col.insert_one(user)
    else:
        updates = {}
        if name and user.get("name") != name:
            updates["name"] = name
        if username and user.get("username") != username:
            updates["username"] = username
        if updates:
            await users_col.update_one({"uid": uid}, {"$set": updates})
    return await users_col.find_one({"uid": uid})

# ======================
# نقاط وفلوس
# ======================
async def get_user_gold(uid):
    user = await get_user(uid)
    return user.get("gold", 0)

async def update_user_gold(uid, amount):
    user = await get_user(uid)
    new_gold = max(0, user.get("gold", 0) + amount)
    await users_col.update_one({"uid": uid}, {"$set": {"gold": new_gold}})
    return new_gold

async def get_user_points(uid):
    user = await get_user(uid)
    return user.get("points", 0)

async def update_user_points(uid, amount):
    user = await get_user(uid)
    new_points = max(0, user.get("points", 0) + amount)
    await users_col.update_one({"uid": uid}, {"$set": {"points": new_points}})
    return new_points

# ======================
# البنك
# ======================
async def deposit_to_bank(uid, amount):
    user = await get_user(uid)
    if amount <= 0 or user.get("gold", 0) < amount:
        return False
    await users_col.update_one({"uid": uid}, {"$inc": {"gold": -amount, "bank": amount}})
    return True

async def withdraw_from_bank(uid, amount):
    user = await get_user(uid)
    if amount <= 0 or user.get("bank", 0) < amount:
        return False
    await users_col.update_one({"uid": uid}, {"$inc": {"bank": -amount, "gold": amount}})
    return True

# ======================
# الرتب
# ======================
async def get_user_rank(uid):
    user = await get_user(uid)
    return user.get("rank", 0)

async def set_user_rank(uid, rank):
    await users_col.update_one({"uid": uid}, {"$set": {"rank": rank}})

# ======================
# الهدايا اليومية
# ======================
DAY = 86400

async def can_take_gift(uid):
    user = await get_user(uid)
    return time.time() - user.get("last_gift_time", 0) >= DAY

async def take_gift(uid, amount=100):
    if not await can_take_gift(uid):
        return None
    await update_user_gold(uid, amount)
    await users_col.update_one({"uid": uid}, {"$set": {"last_gift_time": time.time()}})
    return await get_user_gold(uid)

# ======================
# الرسائل والمستوى
# ======================
async def get_user_message_count(uid):
    user = await get_user(uid)
    return user.get("total_messages", 0)

async def increment_user_message(uid):
    await users_col.update_one({"uid": uid}, {"$inc": {"total_messages": 1}}, upsert=True)

async def get_user_level(uid):
    user = await get_user(uid)
    return user.get("level", 0)

async def set_user_level(uid, level):
    await users_col.update_one({"uid": uid}, {"$set": {"level": level}}, upsert=True)

# ======================
# المجموعات: كتم وطرد
# ======================
async def mute_user(chat_id, user_id):
    group = await groups_col.find_one({"chat_id": chat_id})
    if not group:
        group = {"chat_id": chat_id, "muted": [], "kicked": []}
        await groups_col.insert_one(group)
    if user_id not in group["muted"]:
        group["muted"].append(user_id)
        await groups_col.update_one({"chat_id": chat_id}, {"$set": {"muted": group["muted"]}})

async def unmute_user(chat_id, user_id):
    group = await groups_col.find_one({"chat_id": chat_id})
    if group and user_id in group["muted"]:
        group["muted"].remove(user_id)
        await groups_col.update_one({"chat_id": chat_id}, {"$set": {"muted": group["muted"]}})

async def is_user_muted(chat_id, user_id):
    group = await groups_col.find_one({"chat_id": chat_id})
    return group and user_id in group.get("muted", [])

async def kick_user(chat_id, user_id):
    group = await groups_col.find_one({"chat_id": chat_id})
    if not group:
        group = {"chat_id": chat_id, "muted": [], "kicked": []}
        await groups_col.insert_one(group)
    if user_id not in group["kicked"]:
        group["kicked"].append(user_id)
        await groups_col.update_one({"chat_id": chat_id}, {"$set": {"kicked": group["kicked"]}})

async def get_muted_users(chat_id):
    group = await groups_col.find_one({"chat_id": chat_id})
    return group.get("muted", []) if group else []

async def get_kicked_users(chat_id):
    group = await groups_col.find_one({"chat_id": chat_id})
    return group.get("kicked", []) if group else []

# ======================
# التصدير
# ======================
__all__ = [
    "DEVELOPERS", "get_user", "get_user_gold", "update_user_gold", "get_user_points", "update_user_points",
    "get_user_rank", "set_user_rank",
    "can_take_gift", "take_gift",
    "get_user_message_count", "increment_user_message", "get_user_level", "set_user_level",
    "mute_user", "unmute_user", "is_user_muted", "kick_user",
    "get_muted_users", "get_kicked_users"
]

File: db_manager.js
// db_manager.js
const { MongoClient } = require('mongodb');

const MONGO_URI = "mongodb+srv://wpee923_db_user:08520852KR@cluster0.nzjd5gc.mongodb.net/?retryWrites=true&w=majority";
const DB_NAME = "imperial_bot";

let db, users;

async function initDB() {
    const client = new MongoClient(MONGO_URI);
    try {
        await client.connect();
        db = client.db(DB_NAME);
        users = db.collection("users");
        console.log("✅ MongoDB متصل بنجاح!");
    } catch (err) {
        console.error("❌ فشل الاتصال بـ MongoDB:", err);
    }
}

// ======================
// المطورين
// ======================
const DEVELOPERS = [7076215547, 7855813063, 5860391324];

// ======================
// المستخدمين
// ======================
async function getUser(uid) {
    let user = await users.findOne({ uid });
    if (!user) {
        user = {
            uid,
            gold: 0,
            bank: 0,
            inventory: [],
            rank: 0,
            last_task_time: 0,
            daily_task: null,
            box_ready: false,
            box_opened: false,
            last_gift_time: 0,
            banned: false,
            name: null,
            username: null,
            total_messages: 0,
            married_to: null
        };
        await users.insertOne(user);
    }
    return await users.findOne({ uid });
}

// ======================
// الذهب
// ======================
async function getUserGold(uid) {
    const user = await getUser(uid);
    return user.gold;
}

async function updateUserGold(uid, amount) {
    const user = await getUser(uid);
    const newGold = Math.max(0, user.gold + amount);
    await users.updateOne({ uid }, { $set: { gold: newGold } });
    return newGold;
}

// ======================
// البنك
// ======================
async function getUserBank(uid) {
    const user = await getUser(uid);
    return user.bank;
}

async function depositToBank(uid, amount) {
    const user = await getUser(uid);
    if (amount <= 0 || user.gold < amount) return false;
    await users.updateOne({ uid }, { $inc: { gold: -amount, bank: amount } });
    return true;
}

async function withdrawFromBank(uid, amount) {
    const user = await getUser(uid);
    if (amount <= 0 || user.bank < amount) return false;
    await users.updateOne({ uid }, { $inc: { bank: -amount, gold: amount } });
    return true;
}

// ======================
// المخزون
// ======================
async function getInventory(uid) {
    const user = await getUser(uid);
    return user.inventory || [];
}

async function addToInventory(uid, item, quantity = 1) {
    const user = await getUser(uid);
    const newInv = [...user.inventory, ...Array(quantity).fill(item)];
    await users.updateOne({ uid }, { $set: { inventory: newInv } });
}

async function removeFromInventory(uid, item, quantity = 1) {
    const user = await getUser(uid);
    let count = 0;
    const newInv = [];
    for (let i of user.inventory) {
        if (i === item && count < quantity) {
            count++;
            continue;
        }
        newInv.push(i);
    }
    await users.updateOne({ uid }, { $set: { inventory: newInv } });
    return count === quantity;
}

// ======================
// الرتب
// ======================
async function getUserRank(uid) {
    const user = await getUser(uid);
    return user.rank || 0;
}

async function setUserRank(uid, rank) {
    await users.updateOne({ uid }, { $set: { rank } });
}

async function downgradeUserRank(byUid, targetUid, newRank) {
    if (!DEVELOPERS.includes(byUid)) return { ok: false, error: "❌ هذا الأمر للمطور فقط." };
    if (newRank < 0) newRank = 0;

    const target = await getUser(targetUid);
    const oldRank = target.rank || 0;
    if (newRank >= oldRank) return { ok: false, error: "⚠️ لا يمكن التخفيض لنفس الرتبة أو أعلى." };

    await users.updateOne({ uid: targetUid }, { $set: { rank: newRank } });
    return { ok: true, old_rank: oldRank, new_rank: newRank };
}

// ======================
// المهام اليومية
// ======================
const DAY = 86400 * 1000; // 24 ساعة بالميلي ثانية
const TASKS = [
    { type: "dice", desc: "العب لعبة النرد 🎲" },
    { type: "roulette", desc: "العب روليت 🎰" }
];

async function canGetTask(uid) {
    const user = await getUser(uid);
    return Date.now() - user.last_task_time >= DAY;
}

async function getDailyTask(uid) {
    const user = await getUser(uid);
    if (user.daily_task) return user.daily_task;
    if (!(await canGetTask(uid))) return null;

    const task = TASKS[Math.floor(Math.random() * TASKS.length)];
    await users.updateOne(
        { uid },
        { $set: { daily_task: task, last_task_time: Date.now(), box_ready: false, box_opened: false } }
    );
    return task;
}

async function completeMission(uid, missionType) {
    const user = await getUser(uid);
    if (!user.daily_task || user.daily_task.type !== missionType) return false;
    await users.updateOne({ uid }, { $set: { box_ready: true, daily_task: null } });
    return true;
}

async function canOpenBox(uid) {
    const user = await getUser(uid);
    return user.box_ready && !user.box_opened;
}

async function setBoxOpened(uid) {
    await users.updateOne({ uid }, { $set: { box_opened: true } });
}

// ======================
// الهدايا اليومية
// ======================
async function takeGift(uid, amount = 100) {
    const user = await getUser(uid);
    if (Date.now() - user.last_gift_time < DAY) return null;
    await updateUserGold(uid, amount);
    await users.updateOne({ uid }, { $set: { last_gift_time: Date.now() } });
    return await getUserGold(uid);
}

// ======================
// الحظر
// ======================
async function isUserBanned(uid) {
    const user = await getUser(uid);
    return user.banned || false;
}

async function banUser(uid) {
    await users.updateOne({ uid }, { $set: { banned: true } });
}

async function unbanUser(uid) {
    await users.updateOne({ uid }, { $set: { banned: false } });
}

// ======================
// إحصائيات
// ======================
async function getAllUsersCount() {
    return await users.countDocuments();
}

// ======================
// أعياد الميلاد
// ======================
async function addBirthday(uid, day, month, year = null) {
    if (day < 1 || day > 31 || month < 1 || month > 12) return { ok: false, error: "⚠️ التاريخ غير صالح." };
    await users.updateOne({ uid }, { $set: { birthday: { day, month, year } } });
    return { ok: true, uid, birthday: { day, month, year } };
}

async function removeBirthday(uid) {
    await users.updateOne({ uid }, { $unset: { birthday: "" } });
    return { ok: true, uid };
}

async function getBirthday(uid) {
    const user = await getUser(uid);
    return user.birthday || null;
}

async function listBirthdays() {
    return await users.find({ birthday: { $exists: true } }).toArray();
}

async function enableBirthdayAuto(uid) {
    await users.updateOne({ uid }, { $set: { birthday_auto: true } });
    return { ok: true, uid };
}

async function disableBirthdayAuto(uid) {
    await users.updateOne({ uid }, { $set: { birthday_auto: false } });
    return { ok: true, uid };
}

async function isBirthdayAutoEnabled(uid) {
    const user = await getUser(uid);
    return user.birthday_auto || false;
}

module.exports = {
    initDB,
    DEVELOPERS,
    getUser,
    getUserGold,
    updateUserGold,
    getUserBank,
    depositToBank,
    withdrawFromBank,
    getInventory,
    addToInventory,
    removeFromInventory,
    getUserRank,
    setUserRank,
    downgradeUserRank,
    canGetTask,
    getDailyTask,
    completeMission,
    canOpenBox,
    setBoxOpened,
    takeGift,
    isUserBanned,
    banUser,
    unbanUser,
    getAllUsersCount,
    addBirthday,
    removeBirthday,
    getBirthday,
    listBirthdays,
    enableBirthdayAuto,
    disableBirthdayAuto,
    isBirthdayAutoEnabled
};
